File: Backend_FastAPI/app/schemas/dorm_sync.py
# -*- coding: utf-8 -*-
"""Schema của màn đồng bộ ký túc xá.

🔴 Nguyên tắc xuyên suốt file này: **client không được đặt thứ gì quyết định
phạm vi của một lần ghi**. Lượt đồng bộ hạ được cờ đủ-điều-kiện của cả một khoá
học, nên mọi tham số ảnh hưởng tới "ghi cái gì, cho năm nào" phải do **server**
chốt lúc xem trước và ký trong ``preview_token`` — không phải do request apply
mang lên.

Đó là lý do :class:`DormSyncApplyRequest` chỉ có ĐÚNG một trường và khai
``extra="forbid"``: một request gửi kèm ``academic_year`` phải bị từ chối 422
chứ không được im lặng bỏ qua. Bỏ qua trong im lặng là để client tin rằng mình
đã chọn được năm học — rồi ghi vào một năm khác.
"""

from __future__ import annotations

from typing import List, Optional

from pydantic import BaseModel, ConfigDict, Field


class DormSyncApplyRequest(BaseModel):
    """Thân request của ``POST /apply``. ĐÚNG một trường.

    🔴 ``extra="forbid"``: Pydantic mặc định **bỏ qua** trường lạ. Với endpoint
    này thì bỏ qua là kiểu hỏng tệ nhất — một client (hoặc một bản frontend cũ)
    gửi kèm ``academic_year=2025`` sẽ nhận 200 và tin rằng lượt vừa chạy là cho
    năm 2025, trong khi server ghi vào năm đã ký trong token. Không có gì trên
    màn hình nói ra sự chênh lệch đó.

    ``operation_id`` cũng KHÔNG nằm ở đây: nó do server sinh lúc xem trước và ký
    trong token: nhận nó từ client là mở lại đúng cửa chống-replay mà sổ cái
    ``dorm_sync_operations`` sinh ra để đóng.
    """

    model_config = ConfigDict(extra="forbid")

    preview_token: str = Field(
        ...,
        min_length=1,
        description=(
            "Token server ký ở bước xem trước. Mang theo năm học, operation_id "
            "và dấu vân tay trạng thái mà người bấm đã nhìn."
        ),
    )


class DormSyncPreviewRequest(BaseModel):
    """Thân request của bước xem trước.

    Đây là nơi DUY NHẤT client được chọn năm học — và nó là một lượt **chỉ
    đọc**. Từ đó trở đi năm học đã nằm trong token do server ký.
    """

    model_config = ConfigDict(extra="forbid")

    academic_year: int = Field(
        ...,
        ge=2000,
        le=2100,
        description="Năm học muốn xem trước. Phải là một năm ĐANG MỞ ở hệ KTX.",
    )


class DormSyncContextResponse(BaseModel):
    """Bối cảnh để dựng màn hình: chọn năm nào, mặc định là năm nào.

    ⚠️ ``default_academic_year`` có thể là ``None``. Frontend PHẢI xử lý ca đó
    thay vì tự điền năm hiện tại — "hệ KTX chưa mở năm nào" là một trạng thái
    thật, và đoán một năm ở đây nghĩa là dựng sẵn một lượt ghi vào năm không tồn
    tại bên đích.
    """

    model_config = ConfigDict(extra="forbid")

    open_academic_years: List[int] = Field(
        default_factory=list,
        description="Các năm học đang MỞ ở hệ KTX, sắp giảm dần.",
    )
    default_academic_year: Optional[int] = Field(
        None,
        description=(
            "Năm mở lớn nhất, hoặc ``None`` nếu không có năm nào mở. "
            "KHÔNG bao giờ suy ra từ ngày hệ thống."
        ),
    )


class DormSyncWarningRow(BaseModel):
    """Một người SẮP MẤT CỜ mà vẫn đang giữ giường.

    🔴 Đây là dòng người bấm phải đọc trước khi quyết định. Nó có họ tên và số
    phòng — dữ liệu cá nhân — nên nó đi trong THÂN PHẢN HỒI (sau cổng quyền
    admin), KHÔNG đi trong ``preview_token``: token chỉ được HMAC ký, không
    được mã hoá, và ai cầm chuỗi cũng đọc được.
    """

    model_config = ConfigDict(extra="forbid")

    qlts_profile_id: int
    full_name: str
    building_name: str
    room_code: str
    bed_no: int
    status: str


class DormSyncPreviewResponse(BaseModel):
    """Kết quả bước xem trước. CHỈ ĐỌC — chưa ghi gì sang hệ KTX."""

    model_config = ConfigDict(extra="forbid")

    academic_year: int
    source_count: int = Field(..., description="Số hồ sơ đủ điều kiện ở nguồn QLTS.")

    can_apply: bool = Field(
        ...,
        description=(
            "Có được bấm Ghi hay không. False ⇒ `preview_token` là None và nút "
            "phải bị khoá."
        ),
    )
    blocked_reason: Optional[str] = Field(
        None,
        description="Vì sao chưa ghi được. Chỉ có nghĩa khi `can_apply=False`.",
    )

    warnings: List[DormSyncWarningRow] = Field(
        default_factory=list,
        description="Sắp mất cờ mà vẫn đang giữ giường — người bấm phải đọc.",
    )

    source_hash: Optional[str] = Field(
        None, description="SHA-256 của ảnh chụp nguồn mà người bấm đã xem."
    )
    target_fingerprint: Optional[str] = Field(
        None, description="Dấu vân tay trạng thái chỗ ở phía KTX, do database tính."
    )

    preview_token: Optional[str] = Field(
        None,
        description=(
            "Vé để bấm Ghi. `None` khi chưa ghi được. Mang theo năm học, "
            "operation_id và hai dấu ở trên — KHÔNG mang dữ liệu cá nhân."
        ),
    )
    expires_at: Optional[int] = Field(
        None, description="Thời điểm token hết hạn (epoch giây)."
    )

File: frontend/src/lib/zod/dorm-sync.ts
/**
 * Zod schemas cho màn đồng bộ ký túc xá.
 *
 * Mirrors `Backend_FastAPI/app/schemas/dorm_sync.py` và payload công khai của
 * `DormSyncOperationBlockedError` (`app/utils/exceptions.py`).
 *
 * 🔴 `.strict()` ở MỌI schema. Backend khai `ConfigDict(extra="forbid")` cho
 * cùng những hình dạng này; để frontend nhận thêm trường lạ là mở lại đúng
 * khoảng lệch mà bên kia vừa đóng — và nó sẽ lệch âm thầm, vì một trường thừa
 * không làm gì hỏng cho tới lúc ai đó đọc nhầm nó.
 */
import { z } from "zod"

// =============================================================================
// BỐI CẢNH
// =============================================================================

export const dormSyncContextSchema = z
  .object({
    open_academic_years: z.array(z.number().int()),
    /**
     * ⚠️ `null` là câu trả lời HỢP LỆ: hệ KTX chưa mở năm nào. Giao diện phải
     * xử ca đó chứ KHÔNG tự điền năm hiện tại — điền đại một năm là dựng sẵn
     * một lượt ghi vào năm không tồn tại bên đích.
     */
    default_academic_year: z.number().int().nullable(),
  })
  .strict()

export type DormSyncContext = z.infer<typeof dormSyncContextSchema>

// =============================================================================
// XEM TRƯỚC
// =============================================================================

export const dormSyncWarningSchema = z
  .object({
    qlts_profile_id: z.number().int(),
    full_name: z.string(),
    building_name: z.string(),
    room_code: z.string(),
    bed_no: z.number().int(),
    status: z.string(),
  })
  .strict()

export type DormSyncWarning = z.infer<typeof dormSyncWarningSchema>

export const dormSyncSourceCountsSchema = z
  .object({
    khong_ro_gioi_tinh: z.number().int(),
    chua_chot_nganh: z.number().int(),
    chua_ro_trinh_do: z.number().int(),
    ho_so_dang_xet: z.number().int(),
    khong_co_so_lien_he: z.number().int(),
    co_so_phu: z.number().int(),
    so_bi_bo_vi_qua_dai: z.number().int(),
  })
  .strict()

export type DormSyncSourceCounts = z.infer<typeof dormSyncSourceCountsSchema>

export const dormSyncPreviewSchema = z
  .object({
    academic_year: z.number().int(),
    source_count: z.number().int(),
    /**
     * 🔴 Nút Ghi bám vào ĐÂY, không tự suy từ `source_count > 0`. Backend còn
     * chặn vì những lý do frontend không thấy (năm đã đóng sổ, cấu hình
     * thiếu), và suy lại là dựng một định nghĩa thứ hai.
     */
    can_apply: z.boolean(),
    blocked_reason: z.string().nullable(),
    counts: dormSyncSourceCountsSchema.nullable(),
    warnings: z.array(dormSyncWarningSchema),
    source_hash: z.string().nullable(),
    target_fingerprint: z.string().nullable(),
    snapshot_hash: z.string().nullable(),
    snapshot_version: z.number().int().nullable(),
    /** Vé để bấm Ghi. `null` khi chưa ghi được. KHÔNG giải mã nó ở client. */
    preview_token: z.string().nullable(),
    /** Epoch giây. Dùng TRỰC TIẾP, không đọc `exp` bên trong token. */
    expires_at: z.number().int().nullable(),
  })
  .strict()

export type DormSyncPreview = z.infer<typeof dormSyncPreviewSchema>

// =============================================================================
// GHI
// =============================================================================

export const DORM_SYNC_OUTCOMES = [
  "completed",
  "failed",
  "outcome_unknown",
] as const

export const dormSyncApplySchema = z
  .object({
    operation_id: z.string(),
    academic_year: z.number().int(),
    /** Giao diện rẽ nhánh theo ĐÂY, không theo `message`. */
    outcome: z.enum(DORM_SYNC_OUTCOMES),
    message: z.string(),
    ktx_run_id: z.number().int().nullable(),
    upserted: z.number().int(),
    blocked: z.number().int(),
    deactivated: z.number().int(),
    /**
     * `false` = hệ KTX ĐÃ đổi nhưng sổ đối soát không ghi lại được. Việc đã
     * xảy ra; chỉ nhật ký là thiếu. Tuyệt đối không mời bấm lại ở ca này.
     */
    ledger_saved: z.boolean(),
  })
  .strict()

export type DormSyncApplyResult = z.infer<typeof dormSyncApplySchema>

// =============================================================================
// LỖI CHẶN — payload máy đọc được
// =============================================================================

export const DORM_SYNC_BLOCKED_CODE = "DORM_SYNC_OPERATION_BLOCKED"

export const DORM_SYNC_NEXT_ACTIONS = [
  "wait",
  "preview_again",
  "manual_reconcile",
] as const

export type DormSyncNextAction = (typeof DORM_SYNC_NEXT_ACTIONS)[number]

/**
 * 🔴 Giao diện tin `next_action`, KHÔNG suy từ `operation_status`, không suy
 * từ mã HTTP 409, và tuyệt đối không đọc câu `detail`.
 *
 * `handleApiError` cố ý CHE `detail` của mã `CONFLICT` vì mã đó dùng chung cho
 * nhiều thứ, trong đó có chuỗi nội bộ tiếng Anh. Nếu ta suy trạng thái từ câu
 * chữ thì (a) vi phạm thin-client, (b) vỡ ngay lần backend sửa chính tả, và
 * (c) ca đắt nhất — `manual_reconcile` — sẽ bị đối xử như một lỗi tạm thời.
 */
export const dormSyncBlockedSchema = z
  .object({
    detail: z.string(),
    error_code: z.literal(DORM_SYNC_BLOCKED_CODE),
    operation_status: z.enum(["running", "failed", "outcome_unknown"]),
    next_action: z.enum(DORM_SYNC_NEXT_ACTIONS),
  })
  .strict()

export type DormSyncBlocked = z.infer<typeof dormSyncBlockedSchema>

// =============================================================================
// REQUEST
// =============================================================================

export const dormSyncPreviewRequestSchema = z
  .object({ academic_year: z.number().int().min(2000).max(2100) })
  .strict()

/**
 * ⚠️ ĐÚNG một trường. Backend khai `extra="forbid"`, nên gửi kèm
 * `academic_year` hay `operation_id` sẽ nhận 422 — và đó là chủ ý: năm học và
 * `operation_id` đã nằm trong phiếu do server ký.
 */
export const dormSyncApplyRequestSchema = z
  .object({ preview_token: z.string().min(1) })
  .strict()
